File: carddrawer/src/test/java/com/meli/android/carddrawer/configuration/LightFontConfigurationTest.java
package com.meli.android.carddrawer.configuration;

import androidx.core.content.ContextCompat;
import android.text.TextPaint;

import com.meli.android.carddrawer.BasicRobolectricTest;
import com.meli.android.carddrawer.R;

import com.meli.android.carddrawer.configuration.shadow.ShadowConfiguration;
import com.meli.android.carddrawer.configuration.shadow.ShadowFontConfiguration;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import org.robolectric.RobolectricTestRunner;

import static org.junit.Assert.assertEquals;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyFloat;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

@RunWith(RobolectricTestRunner.class)
public class LightFontConfigurationTest extends BasicRobolectricTest {
    private LightFontConfiguration lightFontConfiguration;

    @Before
    public void doBefore() {
        ShadowConfiguration shadowFontConfiguration = new ShadowFontConfiguration(getContext());
        lightFontConfiguration = new LightFontConfiguration(getContext(), shadowFontConfiguration);
    }

    @Test
    public void getColor_returnsValidColor() {
        int validColor = ContextCompat.getColor(getContext(), R.color.card_drawer_light_font_empty_color);

        assertEquals(validColor, lightFontConfiguration.getColor());
    }

    @Test
    public void setShadow_callSetShadow() {
        TextPaint textPaint = mock(TextPaint.class);

        lightFontConfiguration.setShadow(textPaint);

        verify(textPaint).setShadowLayer(anyFloat(), anyFloat(), anyFloat(), anyInt());
    }

    @Test
    public void setShadow_DoesNotCallSetShadow() {
        LightFontConfiguration lightNotShadowFontConfiguration = new LightFontConfiguration(getContext());
        TextPaint textPaint = mock(TextPaint.class);

        lightNotShadowFontConfiguration.setShadow(textPaint);

        verify(textPaint, never()).setShadowLayer(anyFloat(), anyFloat(), anyFloat(), anyInt());
    }
}

File: carddrawer/src/test/java/com/meli/android/carddrawer/configuration/LightFontConfigurationTest.kt
package com.meli.android.carddrawer.configuration

import android.text.TextPaint
import androidx.core.content.ContextCompat
import com.meli.android.carddrawer.R
import com.meli.android.carddrawer.configuration.base.ConfigurationTestBase
import com.meli.android.carddrawer.configuration.shadow.ShadowConfiguration
import com.meli.android.carddrawer.configuration.shadow.ShadowFontConfiguration
import org.junit.Test
import org.mockito.ArgumentMatchers
import org.mockito.Mockito

class LightFontConfigurationTest: ConfigurationTestBase() {

    private lateinit var lightFontConfiguration: LightFontConfiguration

    private fun initWithShadow() {
        val shadowFontConfiguration: ShadowConfiguration = ShadowFontConfiguration(contextMock)
        lightFontConfiguration = LightFontConfiguration(contextMock, shadowFontConfiguration)
    }

    @Test
    fun `when getting color with shadow font per parameter then it should card_drawer_light_font_empty_color card_drawer_dark_font_empty_color`() {
        initWithShadow()
        val validColor =
            ContextCompat.getColor(contextMock, R.color.card_drawer_light_font_empty_color)
        assert(lightFontConfiguration.color == validColor)
    }

    @Test
    fun `when call function setShadow then it should call setShadowLayer`() {
        initWithShadow()
        val textPaint = Mockito.mock(TextPaint::class.java)
        lightFontConfiguration.setShadow(textPaint)
        Mockito.verify(textPaint).setShadowLayer(
            ArgumentMatchers.anyFloat(),
            ArgumentMatchers.anyFloat(),
            ArgumentMatchers.anyFloat(),
            ArgumentMatchers.anyInt()
        )
    }

    @Test
    fun`when getting color without shadow per parameter then it should return card_drawer_light_font_empty_color`() {
        lightFontConfiguration = LightFontConfiguration(contextMock)
        val validColor =
            ContextCompat.getColor(contextMock, R.color.card_drawer_light_font_empty_color)
        assert(lightFontConfiguration.color == validColor)
    }

    @Test
    fun`when call setShadow without shadow per parameter then it should call setShadowLayer`() {
        val textPaint = Mockito.mock(TextPaint::class.java)
        lightFontConfiguration = LightFontConfiguration(contextMock)
        lightFontConfiguration.setShadow(textPaint)
        Mockito.verify(textPaint, Mockito.never()).setShadowLayer(
            ArgumentMatchers.anyFloat(),
            ArgumentMatchers.anyFloat(),
            ArgumentMatchers.anyFloat(),
            ArgumentMatchers.anyInt()
        )
    }
}